src/main/cpp/src/shuffle_assemble.cu

```diff
@@ -1279,16 +1276,35 @@ __global__ void copy_validity(cudf::device_span<assemble_batch> batches)
   int rows_in_batch =
     min(remaining_rows,
         leading_bytes != 4 ? (leading_bytes * 8) - batch.src_bit_shift : 32 - batch.src_bit_shift);
-  remaining_rows -= rows_in_batch;
 
-  size_type valid_count = 0;
+  // safely load a word, handling alignment and allocation boundaries
+  auto load_word = cuda::proclaim_return_type<bitmask_type>(
+    [] __device__(void const* const _src, int num_bits, int leading_bytes) {
+      uint8_t const* const src_b = reinterpret_cast<uint8_t const*>(_src);
+      if (num_bits > 24) {
+        // if we are aligned we can do a single read. this should be the most common case. only
+        // at the ends of the range of bits to copy will we hit the non-aligned cases.
+        return leading_bytes == 4
+                 ? (reinterpret_cast<bitmask_type const*>(_src))[0]
+                 : static_cast<bitmask_type>(src_b[0]) | static_cast<bitmask_type>(src_b[1] << 8) |
+                     static_cast<bitmask_type>(src_b[2] << 16) |
+                     static_cast<bitmask_type>(src_b[3] << 24);
+      } else if (num_bits > 16) {
```

Comment:
So `leading_bytes == 4` is only relevant for `num_bits > 24`?

Reply:
Correct. It means we've got a full 4, aligned 4 bytes of data at the start of the buffer.